pkg/extensions/sync/sync.go

```diff
@@ -319,186 +326,203 @@ func syncRegistry(ctx context.Context, regCfg RegistryConfig, upstreamURL string
 
 	log.Info().Msgf("got repos: %v", repos)
 
-	var images []struct {
+	upstreamAddr := StripRegistryTransport(upstreamURL)
+
+	reposWithContentID := make(map[string][]struct {
 		ref     types.ImageReference
 		content Content
-	}
-
-	upstreamAddr := StripRegistryTransport(upstreamURL)
+	})
 
 	for contentID, repos := range repos {
 		r := repos
 		contentID := contentID
 
 		if err = retry.RetryIfNecessary(ctx, func() error {
-			refs, err := imagesToCopyFromUpstream(ctx, upstreamAddr, r, upstreamCtx, regCfg.Content[contentID], log)
-			for _, ref := range refs {
-				images = append(images, struct {
-					ref     types.ImageReference
-					content Content
-				}{
-					ref:     ref,
-					content: regCfg.Content[contentID],
-				})
+			for _, repo := range r {
+				refs, err := imagesToCopyFromUpstream(ctx, upstreamAddr, r, upstreamCtx, regCfg.Content[contentID], log)
+				if err != nil {
+					return err
+				}
+
+				for _, ref := range refs[repo] {
+					reposWithContentID[repo] = append(reposWithContentID[repo], struct {
+						ref     types.ImageReference
+						content Content
+					}{
+						ref:     ref,
+						content: regCfg.Content[contentID],
+					})
+				}
 			}
 
-			return err
+			return nil
 		}, retryOptions); err != nil {
 			log.Error().Err(err).Msg("error while getting images references from upstream, retrying...")
 
 			return err
 		}
 	}
 
-	for _, image := range images {
+	for remoteRepo, imageList := range reposWithContentID {
 		select {
 		case <-ctx.Done():
 			return ctx.Err()
 		default:
 			break
 		}
 
-		upstreamImageRef := image.ref
+		remoteRepoCopy := remoteRepo
+		imageStore := storeController.GetImageStore(remoteRepoCopy)
 
-		remoteRepo := getRepoFromRef(upstreamImageRef, upstreamAddr)
-		localRepo := getRepoDestination(remoteRepo, image.content)
-
-		upstreamImageDigest, err := docker.GetDigest(ctx, upstreamCtx, upstreamImageRef)
+		localCachePath, err := getLocalCachePath(imageStore, remoteRepoCopy)
 		if err != nil {
-			log.Error().Err(err).Msgf("couldn't get upstream image %s manifest", upstreamImageRef.DockerReference())
-
-			return err
+			log.Error().Err(err).Msgf("couldn't get localCachePath for %s", remoteRepoCopy)
```

Comment:
why don't you return error here? same with on_demand.go

Reply:
if there is no reason then defer os.RemoveAll() on localCachePath.

Reply:
> why don't you return error here? same with on_demand.go

I must have missed it. Fixed it now, added a return err there

> if there is no reason then defer os.RemoveAll() on localCachePath.

There was a defer os.RemoveAll() on localCachePath, but it was at the end of the big loop that downloads the image, which in hindsight, was not the right place to be anyway. I moved it after the return err above